zebrad/src/commands/start/sync.rs

```diff
@@ -82,62 +91,51 @@ where
-                        "requested more hashes"
-                    );
-
-                    // TODO(jlusby): reject both main and test net genesis blocks
-                    if hashes.last() == Some(&super::GENESIS) {
+                Ok(zn::Response::BlockHeaderHashes(hashes)) => {
+                    if hashes.is_empty() {
+                        tracing::debug!("skipping empty response");
                         continue;
                     }
 
-                    let mut hashes = hashes.into_iter().peekable();
-                    let new_tip = if let Some(tip) = hashes.next() {
-                        tip
-                    } else {
-                        continue;
-                    };
-
                     // ObtainTips Step 3
                     //
                     // For each response, starting from the beginning of the
                     // list, prune any block hashes already included in the
                     // state, stopping at the first unknown hash to get resp1',
                     // ..., respF'. (These lists may be empty).
-                    while let Some(&next) = hashes.peek() {
-                        let resp = self
+                    let mut first_unknown = 0;
+                    for (i, &hash) in hashes.iter().enumerate() {
+                        let depth = self
                             .state
                             .ready_and()
                             .await
                             .map_err(|e| eyre!(e))?
-                            .call(zebra_state::Request::GetDepth { hash: next })
+                            .call(zebra_state::Request::GetDepth { hash })
                             .await
                             .map_err(|e| eyre!(e))?;
-
-                        let should_download = matches!(resp, zebra_state::Response::Depth(None));
-
-                        if should_download {
-                            download_set.extend(hashes);
+                        if let zs::Response::Depth(None) = depth {
+                            first_unknown = i;
                             break;
-                        } else {
-                            let _ = hashes.next();
                         }
                     }
+                    if first_unknown == hashes.len() {
+                        tracing::debug!("no new hashes, even though we gave our tip?");
+                        continue;
+                    }
+                    let unknown_hashes = &hashes[first_unknown..];
+                    download_set.extend(unknown_hashes);
 
                     // ObtainTips Step 4
                     //
                     // Combine the last elements of each list into a set; this
                     // is the set of prospective tips.
+                    let new_tip = *unknown_hashes
+                        .last()
+                        .expect("already checked first_unknown < hashes.len()");
```

Comment:
this is also not necessary if we just use `pop()` on the Vec. I think its cleaner if we just keep the tips and the rest of the unknown_hashes separate.

Reply:
Since `pop` returns an `Option`, there's no way to avoid an `expect`, so I'm not sure what's unnecessary here.  Extracting the new tip has to happen in step 4 *after* pruning the lists in step 3, not before, because we have to check whether we should consider the tip at all.